plugins/c9.ide.ace.keymaps/cli.js

```diff
@@ -187,6 +187,122 @@ define(function(require, exports, module) {
+             */
+            'sed': {
+                regex: /^(%|'<,'>|(?:\d+|\.),(?:\+?\d+|\$|\.))?s(\/|#)(.*?)\2(.*?)\2([giIc]*)$/,
+                action: function (editor, cmd, data) {
+                    var pattern = (
+                        data.match[3]
+                        || (editor.state.cm
+                            && editor.state.cm.$searchHighlight
+                            && editor.state.cm.$searchHighlight.regExp.source)
+                        || '')
+                        .replace(/(\\)?(\(|\)|\+|\?|\||\&|\{)/g, function (m, m1, m2) {
+                            // Revert regular syntax
+                            return m1 ? m2 : '\\' + m2;
+                        });
+                    if (!pattern) {
+                        return editor.cmdLine.setTimedMessage(
+                            'E35: No previous regular expression',
+                            15000);
+                    }
+                    
+                    /**
+                     * g - global by row
+                     * i - ignorecase
+                     * I - noignorecase
+                     * todo: c - confirm
+                     */
+                    var flags = (data.match[5] || '').replace(/[cI]/g, '');
+                    var re = new RegExp(pattern, flags);
+                    
+                    var replacement = data.match[4] || '';
+                    var text = editor.getValue();
+                    var parts = text.split('\n');
+                    var range = getCmdLineRange(editor, data.match[1], parts);
+                    var lastReplaceRange;
+                    var lineCount = 0;
+                    var subCount = 0;
+                    
+                    for (var i = range[0]; i <= range[1]; i++) {
+                        var isChanged = false;
+                        parts[i] = parts[i].replace(re, function () {
+                            var args = [].slice.call(arguments, 0);
+                            var offset = args[args.length - 2];
+                            if (!isChanged) {
+                                lineCount += 1;
+                                isChanged = true;
+                            }
+                            
+                            subCount += 1;
+                            lastReplaceRange = {
+                                'row': i,
+                                'column': offset
+                            };
+                            
+                            return replacement.replace(/\\(\d+)/g, function (m, m1) {
+                                var i = parseInt(m1, 10);
+                                return i < args.length - 2 ? args[i] : '';
+                            });
+                        });
+                    }
+                    
+                    if (lastReplaceRange) {
+                        editor.setValue(parts.join('\n'), -1);
```

Comment:
replacing this way will cause problems with undo history and selection. I think we need to either use the api from https://github.com/ajaxorg/ace/blob/master/lib/ace/search.js or https://github.com/c9/core/blob/master/plugins/c9.ide.find.replace/findreplace.js#L983